### PR TITLE
Create a collxml parser using SAX library

### PR DESCRIPTION
**Pull request by BryanEliDimas**
For issue https://github.com/Connexions/cnx-press/issues/162 Implement a function to parse collxml to a tree structure

### CODE REVIEW

File: press/models.py [hunk 1]
```diff
@@ -25,3 +25,61 @@
...
+        """Equality is defined as two collections having the same
+        (type of) elements in the same order and all modules having
+        the same title.
+        """
+        if self.name == 'title' and other.name == 'title':
+            # title tags may have nested tags within them,
+            # so consider the text in those as well.
+            return self._complete_title() == other._complete_title()
+        else:
+            return self.name == other.name
```
Comment (pumazi) [inline press/models.py]: Lookup special method `__eq__`, so that `tree == tree` works.

File: press/utils.py [hunk 2]
```diff
@@ -2,9 +2,20 @@
     'convert_to_legacy_domain',
     'convert_version_tuple_to_version_string',
     'convert_version_to_legacy_version',
+    'requires_major_version_update',
 )
 
 
+def requires_major_version_update(before, after):
```
Comment (pumazi) [inline press/utils.py]: Bit of scope creep, but a good addition none the less. Though, I'd have liked to see this as a stand alone PR.

File: tests/unit/parsers/test_collection.py [hunk 3]
```diff
@@ -4,7 +4,43 @@
...
+    """
+    titles_with_markup = [
+        'Introduction to Quartus and Circuit Diagram Design SOME MATH',
+        'Lab 1-1: 4-Bit Mux and all NAND/NOR Mux SOME STYLED TEXT',
+        'Lab 4-1 Interrupt Driven Programming in MSP430 Assembly '
+        'SOME DIV TAG SOME SPAN TAG',
+    ]
+
+    # TODO: ideally, the text that gets parsed should be the following
+    # (but it's tricky to do with a SAX parser):
```
Comment (pumazi) [inline tests/unit/parsers/test_collection.py]: If this is really an issue you could look into lxml's parser, which I unfortunately may not have mentioned. https://lxml.de/parsing.html#iterparse-and-iterwalk It could help by allowing you to subscribe to specific elements and give you access to lxml.Element objects, which have an easier API to work with.

File: press/models.py (hunk 1 above)
Comment (BryanEliDimas) [reply press/models.py]: I had implemented that method, actually, but I received very strong pushback against it from @kevinburleigh75—basically because it may cause more trouble in the future than it's worth, and @reedstrm seemed to agree. I considered, however, implementing it in a separate object/data structure whose sole purpose is to be used for comparing *and* diffying two trees, but I didn't get to that point yet, not in this PR at least. That specialized tree structure would also implement `__hash__` in addition to `__eq__` (https://docs.python.org/3.7/reference/datamodel.html#object.__hash__).

File: press/utils.py (hunk 2 above)
Comment (BryanEliDimas) [reply press/utils.py]: I see! Agreed; although, it came in very handy as a means to test that the rest of my code was working properly, for this same PR.

File: tests/unit/parsers/test_collection.py (hunk 3 above)
Comment (BryanEliDimas) [reply tests/unit/parsers/test_collection.py]: That's interesting/useful. However, the problem isn't exactly with the SAX parser but with any stream parser, or more specifically, with trailing text using a stream parser.
This should hopefully demo the problem... for a title tag:
`<title>This is text <sub>and</sub>this is trailing text.</title>`

```
Start: This is text
Start: and
Start: this is trailing text.
```

Everything I tried gave me either this:
`This is text this is trailing text and`
Or this:
`and This is text this is trailing text`

It sort of occurred to me one way to parse the text correctly as part of the title tag is to be able to tell when I'm inside a title tag and any text after that point gets parsed as part of the tile but what would I do with the `sub` tag? Pretend it doesn't exist (ignore it completely and any others inside it)? Leave it alone but still pretend its text is actually directly within the `title` tag _somehow_?

Please, if you have any ideas, or any idea what I mean (I think I did a terrible job at explaining the problem)—let me know! @pumazi

**Comment by BryanEliDimas** [comment]
Excellent! 😃Thanks for the code review @pumazi! I addressed your comments, and I just have one really important question: what do you mean by "exposing the xml in the end result"?

**Comment by BryanEliDimas** [comment]
Seems like I'm on track with this PR... moving this PR to code review. cc @reedstrm

**Comment by BryanEliDimas** [comment]
Per Michael Mulich's code review, I am working on removing the XML files from the testing data directory. I'm considering moving them to the litezip library, as Ross has suggested to me, but I'm also considering adding in a "collxml" file parsed as JSON as that is much much easier to work with than the existing code that uses Jinja templates to create and manipulate xml.

**Comment by BryanEliDimas** [comment]
Hey @pumazi, may I please know the exact reason why I should remove the XML template files from the testing directory? I see that _some_ similar files exist in cnx-litezip, and they are used in this repo as fixtures. Could I do the same with this PR?

**Comment by brittweinstein** [comment]
We are not sure of the status of this item. Ross demoed this on Friday however this PR is still open.

File: press/errors.py [hunk 4]
```diff
@@ -12,7 +12,5 @@ def __init__(self, model):
         self.model = model
 
 
-class CollectionChanged(Exception):
-    """Raised when checked out version is older than published version"""
-    def __init__(self, collection):
-        self.collection = collection
+class AttributeDoesNotExist(Exception):
```
Comment (BryanEliDimas) [inline press/errors.py]: Btw, I am totally open to suggestions on a better name for this or if this is even necessary.

File: press/legacy_publishing/litezip.py [hunk 5]
```diff
@@ -34,13 +34,6 @@ def publish_litezip(struct, submission, db_conn, coll_changes_allowed=True):
     except IndexError:  # pragma: no cover
         raise NotImplementedError('litezip without collection')
 
-    try:
```
Comment (BryanEliDimas) [inline press/legacy_publishing/litezip.py]: I have no idea how this code got into master previously. It's clearly bogus, so I'm just deleting it here which would fix issue #308

Comment (pumazi) [reply press/legacy_publishing/litezip.py]: This is to allow for collection structure only publications. That is if you send a litezip that only contains a `collection.xml`.

**Comment by pumazi** [comment]
This PR could be smaller, no? There are at least two different things happening in this PR.

IIRC the original task was to create a parser for a collection tree. That would have been a good size PR with a very focused set of logic. Another PR could have been made to integrate that logic into the publication procedure. The separation is therefore clear that we have a parser and we have logic that uses that parser.

File: press/legacy_publishing/collection.py [hunk 6]
```diff
@@ -55,17 +58,36 @@ def publish_legacy_book(model, metadata, submission, db_conn, changed=None):
...
-    for res in model.resources:
-        if res.sha1 != existing_shas.get(res.filename):
-            raise CollectionChanged(model)
+    if changed is False and existing_shas['collection.xml'] \
+            == produce_hashes_from_filepath(model.file)['sha1']:
+        for res in model.resources:
+            if res.sha1 != existing_shas[res.filename]:
+                break  # publish!
+        else:  # collxml and all resources are identical to already published
+            raise Unchanged(model)
```
Comment (pumazi) [inline press/legacy_publishing/collection.py]: Looks like this `CollectionChanged` and `Unchanged` code is already in, so my comments are just comments...

This use of state as exceptions smells funny. Just pointing that out. 😸

**Comment by pumazi** [comment]
I will state that I'm concerned that if the renaming and abstracting of the `PressElement` logic doesn't happen now, it will be slightly more difficult to change later, because this type of code has a way of meandering into strange places. This inevitably causes the class API to lock into place.
Hopefully that won't happen. I feel better having said it at least. 😸

**Comment by BryanEliDimas** [comment]
@pumazi re: "My expectations were to have an implementation of a Tree model (and siblings) that is not coupled to XML."
I understand now, but I didn't when I started working on it as it wasn't listed as a requirement from the beginning. The conversation which made it clear happened in Slack DMs on December 14.

File: press/legacy_publishing/litezip.py [hunk 7]
```diff
@@ -67,8 +61,8 @@ def publish_litezip(struct, submission, db_conn):
         except Unchanged:
             pass  # only publish content that has changed.
 
-    any_changes = id_map != {}
-    if any_changes:
+    collxml_changed = bool(id_map)
+    if id_map:
```
Comment (BryanEliDimas) [inline press/legacy_publishing/litezip.py]: And if you rename the line above, you can use it here. 
`if modules_changed:  # update the collection`

File: tests/functional/legacy_publishing/test_litezip.py [hunk 8]
```diff
@@ -65,3 +65,47 @@ def test_publish_litezip(
...
+    collection, tree, modules = content_util.rebuild_collection(collection,
+                                                                tree)
+    collection = persist_util.insert_collection(collection)
+
+    # ... remove second element from the tree ...
+    tree.pop(1)
+
+    collection, tree, modules = content_util.rebuild_collection(collection,
+                                                                tree)
+    struct = tuple([collection])
```
Comment (BryanEliDimas) [inline tests/functional/legacy_publishing/test_litezip.py]: I like how Mulich adds a comment to denote what is the important part of a test, so do it above this line: add a comment `#  TARGET`

File: press/legacy_publishing/utils.py [hunk 9]
```diff
@@ -103,6 +110,12 @@ def roles_as_dict(roles):
     return authors_dict
 
 
+def versions_as_set(modules):
```
Comment (BryanEliDimas) [inline press/legacy_publishing/utils.py]: OH, nice!!

File: press/legacy_publishing/litezip.py [hunk 10]
```diff
@@ -67,8 +61,8 @@ def publish_litezip(struct, submission, db_conn):
         except Unchanged:
             pass  # only publish content that has changed.
 
-    any_changes = id_map != {}
-    if any_changes:
+    collxml_changed = bool(id_map)
```
Comment (BryanEliDimas) [inline press/legacy_publishing/litezip.py]: Nice! bool(id_map) is much cleaner than what I had. But, an empty id_map actually means that no modules changed, so I'd suggest renaming collxml_changed to modules_changed.

File: press/legacy_publishing/utils.py (hunk 9 above)
Comment (BryanEliDimas) [inline press/legacy_publishing/utils.py]: OH, nice!!

File: press/legacy_publishing/litezip.py (hunk 10 above)
Comment (karenc) [reply press/legacy_publishing/litezip.py]: Yes, I think that makes sense.  I'll change it. :+1:

**Comment by karenc** [comment]
I've split this PR up into #329, #333 and #334.

You can see the difference between this PR and the three PRs by doing:

```
git diff collection_xml_parsing_util..major-minor-rev
```
